### ISSUE TITLE
g2o Memory leak

### ISSUE THREAD

**Issue by anson2004110** [RainerKuemmerle/g2o#167]
there are many g2o instances in the demo cpp for memory using, for example
in the g2o/g2o/examples/sba/sba_demo.cpp， it shows that:
g2o::OptimizationAlgorithmLevenberg* solver = new g2o::OptimizationAlgorithmLevenberg(solver_ptr);
but we cannot find any delete memory command at the end of codes. So my question is how to release these memory usage？

**Comment by WorldEditors** [comment]
I used to have the similar doubt like you until I find the answer: g2o actually deletes every thing from single one entrance.

Here is a example:   "SparseOptimizer" class, in "sparse_optimizer.cpp", it writes: 

  SparseOptimizer::~SparseOptimizer(){
    delete _algorithm;
    G2OBatchStatistics::setGlobalStats(0);
  }
which deletes the algorithm class, while in "optimization_algorithm_with_hessian", it writes

OptimizationAlgorithmWithHessian::~OptimizationAlgorithmWithHessian()
  {
    delete _solver;
  }

so basically you can see that in G2O demo it news a lot of things without delete, the key is that it deletes everything start from SparseOptimizer class. I think the initiative for this is to avoid the verbose delete process.

However, from the stance of the stability of code, I do think it is a bad idea. For if I defined a SparseOptimizer without using  setAlgorithm(), the destruct function will have severe problem
